File: 2-Graph Search, Shortest Path, and Data Structures/2-Binary Search Tree/3-Red-Black Tree/RedBlackTree.java
/**
 * Invariants of a Red-Black Tree:
 * 1. Each node is either red or black.
 * 2. The root is always black.
 * 3. Never allow two red nodes in a row
 * 4. Every root-null path passes exactly the same number of black nodes.
 *    i.e., The number of black nodes that a root-null passes cannot depend on
 *          the path.
 *
 * Implementation according to:
 * https://zhuanlan.zhihu.com/p/37470948
 *
 * @author Ziang Lu
 */
public class RedBlackTree {

    /**
     * Static nested Node class.
     */
    private static class Node {
        /**
         * Key of this node.
         */
        private int key;
        /**
         * Reference to the parent.
         */
        private Node parent;
        /**
         * Reference to the left child.
         */
        private Node left;
        /**
         * Reference to the right child.
         */
        private Node right;
        /**
         * Color of this node.
         */
        private boolean color;

        /**
         * Constructor with parameter.
         * @param key key to store
         * @param color color of this node
         */
        Node(int key, boolean color) {
            this.key = key;
            parent = null;
            left = null;
            right = null;
            this.color = color;
        }

        @Override
        public String toString() {
            StringBuilder s = new StringBuilder();
            s.append("[").append(key).append(" ");
            if (color == RED) {
                s.append("R");
            } else {
                s.append("B");
            }
            s.append("]");
            return s.toString();
        }
    }

    /**
     * Color red.
     */
    private static final boolean RED = true;
    /**
     * Color black.
     */
    private static final boolean BLACK = false;

    /**
     * Root of this Red-Black Tree.
     */
    private Node root;

    /**
     * Default constructor.
     */
    public RedBlackTree() {
        root = null;
    }

    /**
     * Searches for the given key in the Red-Black Tree.
     * @param key key to search for
     * @return whether the key is in the Red-Black Tree
     */
    public boolean search(int key) {
        return searchHelper(key, root);
    }

    /**
     * Private helper method to search for the given key in the given subtree
     * recursively.
     * @param key key to search for
     * @param curr current node
     * @return whether the key is in the subtree
     */
    private boolean searchHelper(int key, Node curr) {
        // Base case 1: Not found
        if (curr == null) {
            return false;
        }
        // Base case 2: Found it
        if (curr.key == key) {
            return true;
        }

        // Recursive case
        if (curr.key > key) {
            return searchHelper(key, curr.left);
        } else {
            return searchHelper(key, curr.right);
        }
    }

    /**
     * Insertion:
     * 1. Insert a new element as a normal BST (Might break invariants)
     *    Note that the newly inserted node is always red, since the probability
     *    of breaking the invariants for a red node is smaller than that of a
     *    black node
     *    i.e., If we insert a red node, there is a half-half probability of
     *          breaking invariant #3; if we insert a black node, we must have
     *          broken invariant #4, which is also more difficult to restore.
     * 2. Recolor and perform rotations until invariants are restored
     * @param key key to insert
     */
    public void insert(int key) {
        insertHelper(key, null, root, true);
    }

    /**
     * Private helper method to insert the given key to the given subtree
     * recursively.
     * @param key key to insert
     * @param curr current node
     * @param isLC whether the current node is the left child
     */
    private void insertHelper(int key, Node parent, Node curr, boolean isLC) {
        // Base case 1: Found the spot to insert
        if (curr == null) {
            Node newNode = new Node(key, RED);
            newNode.parent = parent;
            if (parent == null) {
                root = newNode;
            } else {
                if (isLC) {
                    parent.left = newNode;
                } else {
                    parent.right = newNode;
                }
            }

            // Restore the invariants   [O(log n)]
            restoreInvariants(newNode);
            return;
        }
        // Base case 2: Found it
        if (curr.key == key) {
            // No duplicate allowed
            return;
        }

        // Recursive case
        if (curr.key > key) {
            insertHelper(key, curr, curr.left, true);
        } else {
            insertHelper(key, curr, curr.right, false);
        }
        // T(n) = T(n/2) + O(1)
        // a = 1, b = 2, d = 0
        // According to Master Method, the overall running time complexity is O(log n).
    }

    /**
     * Helper method to restore the invariants.
     * @param toRestore node to restore (must be red)
     */
    private void restoreInvariants(Node toRestore) {
        // Case 1: Insert the first node (as the root)
        // Since toRestore is red, this only violates invariant #2.
        // We simply need to recolor toRestore to black.
        if (toRestore == root) {
            toRestore.color = BLACK;
            return;
        }

        // Case 2: The parent is black.
        // This won't violate invariant #3.

        // Case 3: The parent is red.
        // This violates invariant #3.
        Node parent = toRestore.parent;
        if (parent.color == RED) {
            // Then toRestore must have a grandparent, and it must be black according to invariant #3.
            Node grandparent = toRestore.parent.parent;

            if (parent == grandparent.left) {
                Node uncle = grandparent.right;

                if ((uncle != null) && (uncle.color == RED)) { // Case 3.1: The uncle is red.
                    // To restore invariant #3 and #4, simply recolor the grandparent to red and the parent and the uncle to
                    // black.
                    restoreParentAndUncleAreRed(grandparent, parent, uncle);

                    // In this way, the same case applies to the newly recolored grandparent.
                    // Then we simply need to recurse towards grandparent.
                    restoreInvariants(grandparent);
                } else { // Case 3.2: The uncle is black.
                    if (toRestore == parent.right) { // Case 3.2.1: toRestore is the right child.
                        /*
                                 Black
                                /   \
                               Red   Black
                                 \
                               Red (Curr)
                         */
                        leftRotate(parent);

                        // In this way, the same case applies to the newly left rotated parent.
                        // Then we simply need to recurse towards parent.
                        restoreInvariants(parent);
                    } else { // Case 3.2.2: toRestore is the left child.
                        /*
                                 Black
                                 /   \
                               Red   Black
                              /
                            Red (Curr)
                         */
                        // Recolor the parent to black, and the grandparent to red
                        parent.color = BLACK;
                        grandparent.color = RED;
                        rightRotate(grandparent);
                    }
                }
            } else {
                Node uncle = grandparent.left;

                if ((uncle != null) && (uncle.color == RED)) { // Case 3.1: The uncle is red.
                    restoreParentAndUncleAreRed(grandparent, parent, uncle);
                    restoreInvariants(grandparent);
                } else { // Case 3.2: The uncle is black.
                    if (toRestore == parent.left) { // Case 3.2.1: toRestore is the left child.
                        /*
                                 Black
                                 /   \
                              Black   Red
                                     /
                                Red (Curr)
                         */
                        rightRotate(parent);

                        // In this way, the same case applies to the newly left rotated parent.
                        // Then we simply need to recurse towards parent.
                        restoreInvariants(parent);
                    } else { // Case 3.2.2: toRestore is the right child.
                        /*
                                 Black
                                /   \
                             Black   Red
                                       \
                                    Red (Curr)
                         */

                        // Recolor the parent to black, and the grandparent to red
                        parent.color = BLACK;
                        grandparent.color = RED;
                        leftRotate(grandparent);
                    }
                }
            }
        }

        // Recolor the root to black to ensure invariant #2
        root.color = BLACK;
        // Running time complexity: O(log n)
    }

    /**
     * Helper method to restore the invariants when the parent and the uncle are
     * both red.
     * @param grandparent grandparent
     * @param parent parent node
     * @param uncle uncle node
     */
    private void restoreParentAndUncleAreRed(Node grandparent, Node parent, Node uncle) {
        grandparent.color = RED;
        parent.color = BLACK;
        uncle.color = BLACK;
        // Running time complexity: O(1)
    }

    /**
     * Helper method to do a left rotation on the given node.
     * @param toRotate node to rotate on
     */
    private void leftRotate(Node toRotate) {
        // Temporarily store the parent and the right child
        Node parent = toRotate.parent;
        Node tmp = toRotate.right;

        // Reconnect the references to realize left rotation
        toRotate.right = tmp.left;
        if (tmp.left != null) {
            tmp.left.parent = toRotate;
        }

        tmp.left = toRotate;
        toRotate.parent = tmp;

        if (parent != null) {
            if (toRotate == parent.left) {
                parent.left = tmp;
            } else {
                parent.right = tmp;
            }
        }
        tmp.parent = parent;

        if (toRotate == root) {
            root = tmp;
        }
        // Running time complexity: O(1)
    }

    /**
     * Helper method to do a right rotation on the given node.
     * @param toRotate node to rotate on
     */
    private void rightRotate(Node toRotate) {
        // Temporarily store the parent and the left child
        Node parent = toRotate.parent;
        Node tmp = toRotate.left;

        // Reconnect the references to realize right rotation
        toRotate.left = tmp.right;
        if (tmp.right != null) {
            tmp.right.parent = toRotate;
        }

        tmp.right = toRotate;
        toRotate.parent = tmp;

        if (parent != null) {
            if (toRotate == parent.left) {
                parent.left = tmp;
            } else {
                parent.right = tmp;
            }
        }
        tmp.parent = parent;

        if (toRotate == root) {
            root = tmp;
        }
        // Running time complexity: O(1)
    }

    /**
     * Traverses this Red-Black Tree in-order.
     */
    public void traverseInOrder() {
        StringBuilder s = new StringBuilder();
        traverseInOrderHelper(root, s);
        System.out.println(s.toString().trim());
    }

    /**
     * Private helper method to traverse the given subtree in-order recursively.
     * @param curr current node
     * @param s StringBuilder to append
     */
    private void traverseInOrderHelper(Node curr, StringBuilder s) {
        // Base case
        if (curr == null) {
            return;
        }
        // Recursive case
        traverseInOrderHelper(curr.left, s);
        s.append(curr.toString()).append(' ');
        traverseInOrderHelper(curr.right, s);
    }

}

File: 2-Graph Search, Shortest Path, and Data Structures/2-Binary Search Tree/3-Red-Black Tree/red_black_tree.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Invariants of a Red-Black Tree:
1. Each node is either red or black.
2. The root is always black.
3. Never allow two red nodes in a row
4. Every root-null path passes exactly the same number of black nodes.
   i.e., The number of black nodes that a root-null passes cannot depend on
         the path.

Implementation according to:
https://zhuanlan.zhihu.com/p/37470948
"""

__author__ = 'Ziang Lu'


class Node(object):
    __slots__ = ['_key', '_parent', '_left', '_right', '_color']

    def __init__(self, key: int, color: bool):
        """
        Constructor with parameter.
        :param key: int
        :param color: boolean
        """
        self._key = key
        self._parent = None
        self._left = None
        self._right = None
        self._color = color

    @property
    def key(self) -> int:
        """
        Accessor of key.
        :return: int
        """
        return self._key

    @property
    def parent(self):
        """
        Accessor of parent.
        :return: Node
        """
        return self._parent

    @property
    def left(self):
        """
        Accessor of left.
        :return: Node
        """
        return self._left

    @property
    def right(self):
        """
        Accessor of right.
        :return: Node
        """
        return self._right

    @property
    def color(self) -> bool:
        """
        Accessor of color.
        :return: None
        """
        return self._color

    @parent.setter
    def parent(self, parent) -> None:
        """
        Mutator of parent.
        :param parent: Node
        :return: None
        """
        self._parent = parent

    @left.setter
    def left(self, left) -> None:
        """
        Mutator of left.
        :param left: Node
        :return: None
        """
        self._left = left

    @right.setter
    def right(self, right) -> None:
        """
        Mutator of right.
        :param right: Node
        :return: None
        """
        self._right = right

    @color.setter
    def color(self, color: bool) -> None:
        """
        Mutator of color.
        :param color: boolean
        :return: None
        """
        self._color = color

    def __repr__(self):
        s = f'[{self._key} '
        if self._color == RedBlackTree.RED:
            s += 'R'
        else:
            s += 'B'
        s += ']'
        return s


class RedBlackTree(object):
    __slots__ = ['_root']

    RED = True
    BLACK = False

    def __init__(self):
        """
        Default constructor.
        """
        self._root = None

    def search(self, key: int) -> bool:
        """
        Searches for the given key in the Red-Black Tree.
        :param key: int
        :return: boolean
        """
        return self._search_helper(key, self._root)

    def _search_helper(self, key: int, curr: Node) -> bool:
        """
        Private helper method to search for the given key in the given subtree
        recursively.
        :param key: int
        :param curr: Node
        :return: boolean
        """
        # Base case 1: Not found
        if not curr:
            return False
        # Base case 2: Found it
        if curr.key == key:
            return True

        # Recursive case
        if curr.key > key:
            return self._search_helper(key, curr.left)
        else:
            return self._search_helper(key, curr.right)

    def insert(self, key: int) -> None:
        """
        Insertion:
        1. Insert a new element as a normal BST (Might break invariants)
           Note that the newly inserted node is always red, since the
           probability of breaking the invariants for a red node is smaller than
           that of a black node
           i.e., If we insert a red node, there is a half-half probability of
                 breaking invariant #3; if we insert a black node, we must have
                 broken invariant #4, which is also more difficult to restore.
        2. Recolor and perform rotations until invariants are restored
        :param key: int
        :return: None
        """
        self._insert_helper(key, parent=None, curr=self._root, is_lc=True)

    def _insert_helper(self, key: int, parent: Node, curr: Node,
                       is_lc: bool) -> None:
        """
        Private helper method to insert the given key to the given subtree
        recursively.
        :param key: int
        :param parent: Node
        :param curr: Node
        :param is_lc: boolean
        :return: None
        """
        # Base case 1: Found the spot to insert
        if not curr:
            new_node = Node(key, self.RED)
            new_node.parent = parent
            if not parent:
                self._root = new_node
            else:
                if is_lc:
                    parent.left = new_node
                else:
                    parent.right = new_node

            # Restore the invariants   [O(log n)]
            self._restore_invariants(to_restore=new_node)
            return
        # Base case 2: Found it
        if curr.key == key:
            # No duplicate allowed
            return

        # Recursive case
        if curr.key > key:
            self._insert_helper(key, parent=curr, curr=curr.left, is_lc=True)
        else:
            self._insert_helper(key, parent=curr, curr=curr.right, is_lc=False)
        # T(n) = T(n/2) + O(1)
        # a = 1, b = 2, d = 0
        # According to Master Method, the overall running time complexity is
        # O(log n).

    def _restore_invariants(self, to_restore: Node) -> None:
        """
        Helper method to restore the invariants.
        :param to_restore: Node
        :return: None
        """
        # Case 1: Insert the first node (as the root)
        # Since to_store is red, this only violates invariant #2.
        # We simply need to recolor toRestore to black.
        if to_restore is self._root:
            to_restore.color = self.BLACK
            return

        # Case 2: The parent is black.
        # This won't violate invariant #3.

        # Case 3: The parent is red.
        # This violates invariant #3.
        parent = to_restore.parent
        if parent.color == self.RED:
            # Then to_restore must have a grandparent, and it must be black
            # according to invariant #3.
            grandparent = to_restore.parent.parent

            if parent is grandparent.left:
                uncle = grandparent.right

                if uncle and uncle.color == self.RED:
                    # Case 3.1: The uncle is red.
                    #         Black
                    #        /    \
                    #     Red     Red
                    #     /
                    # Red (Curr)
                    # To restore invariant #3 and #4, simply recolor the
                    # grandparent to red and the parent and the uncle to black.
                    self._restore_parent_and_uncle_are_red(grandparent, parent,
                                                           uncle)
                    #          Red
                    #        /    \
                    #     Black  Black
                    #     /
                    # Red (Curr)

                    # In this way, the same case applies to the newly recolored
                    # grandparent.
                    # Then we simply need to recurse towards grandparent.
                    self._restore_invariants(to_restore=grandparent)
                else:
                    # Case 3.2: The uncle is black.
                    if to_restore is parent.right:
                        # Case 3.2.1: to_restore is the right child.
                        #         Black
                        #        /    \
                        #     Red    Black
                        #       \
                        #    Red (Curr)
                        self._left_rotate(to_rotate=parent)

                        # In this way, the same case applies to the newly left
                        # rotated parent.
                        # Then we simply need to recurse towards parent.
                        self._restore_invariants(to_restore=parent)
                    else:
                        # Case 3.2.2: to_restore is the left child.
                        #         Black
                        #        /    \
                        #     Red    Black
                        #     /
                        # Red (Curr)
                        # Recolor the parent to black, and the grandparent to
                        # red
                        parent.color = self.BLACK
                        grandparent.color = self.RED
                        self._right_rotate(to_rotate=grandparent)
            else:
                uncle = grandparent.left

                if uncle and uncle.color == self.RED:
                    # Case 3.1: The uncle is red.
                    self._restore_parent_and_uncle_are_red(grandparent, parent,
                                                           uncle)
                    self._restore_invariants(to_restore=grandparent)
                else:
                    # Case 3.2: The uncle is black.
                    if to_restore is parent.left:
                        # Case 3.2.1: to_restore is the left child.
                        #     Black
                        #    /    \
                        # Black   Red
                        #         /
                        #    Red (Curr)
                        self._right_rotate(to_rotate=parent)

                        # In this way, the same case applies to the newly  left
                        # rotated parent.
                        # Then we simply need to recurse towards parent.
                        self._restore_invariants(to_restore=parent)
                    else:
                        # Case 3.2.2: to_restore is the right child
                        #     Black
                        #    /    \
                        # Black   Red
                        #           \
                        #        Red (Curr)

                        # Recolor the parent to black, and the grandparent to
                        # red
                        parent.color = self.BLACK
                        grandparent.color = self.RED
                        self._left_rotate(to_rotate=grandparent)

        # Recolor the root to black to ensure invariant #2
        self._root.color = self.BLACK
        # Running time complexity: O(log n)

    def _restore_parent_and_uncle_are_red(self, grandparent: Node, parent: Node,
                                          uncle: Node) -> None:
        """
        Helper method to restore the invariants when the parent and the uncle
        are both red.
        :param grandparent: Node
        :param parent: Node
        :param uncle: Node
        :return: None
        """
        grandparent.color = self.RED
        parent.color = self.BLACK
        uncle.color = self.BLACK
        # Running time complexity: O(1)

    def _left_rotate(self, to_rotate: Node) -> None:
        """
        Helper method to do a left rotation on the given node.
        :param to_rotate: Node
        :return: None
        """
        # Temporarily store the parent and the right child
        parent = to_rotate.parent
        tmp = to_rotate.right

        # Reconnect the references to realize left rotation
        to_rotate.right = tmp.left
        if tmp.left:
            tmp.left.parent = to_rotate

        tmp.left = to_rotate
        to_rotate.parent = tmp

        if parent:
            if to_rotate is parent.left:
                parent.left = tmp
            else:
                parent.right = tmp
        tmp.parent = parent

        if to_rotate is self._root:
            self._root = to_rotate
        # Running time complexity: O(1)

    def _right_rotate(self, to_rotate: Node) -> None:
        """
        Helper method to do a right rotation on the given node.
        :param to_rotate: Node
        :return: None
        """
        # Temporarily store the parent and the left child
        parent = to_rotate.parent
        tmp = to_rotate.left

        # Reconnect the references to realize right rotation
        to_rotate.left = tmp.right
        if tmp.right:
            tmp.right.parent = to_rotate

        tmp.right = to_rotate
        to_rotate.parent = tmp

        if parent:
            if to_rotate is parent.left:
                parent.left = tmp
            else:
                parent.right = tmp
        tmp.parent = parent

        if to_rotate is self._root:
            self._root = to_rotate
        # Running time complexity: O(1)

    def traverse_in_order(self) -> None:
        """
        Traverses the Red-Black Tree in-order.
        :return: None
        """
        s = self._traverse_in_order_helper(self._root)
        print(s.strip())

    def _traverse_in_order_helper(self, curr: Node) -> str:
        """
        Private helper function to traverse the given subtree in-order
        recursively.
        :param curr: Node
        :return: str
        """
        # Base case
        if not curr:
            return ''
        # Recursive case
        s = self._traverse_in_order_helper(curr.left)
        s += str(curr) + ' '
        s += self._traverse_in_order_helper(curr.right)
        return s
